### ISSUE TITLE
Add a flag to always use payloads from builders

### ISSUE THREAD

**Issue by paulhauner** [sigp/lighthouse#4040]
## Description

In #3934 we added support for the new payload value APIs, which means we've started using local payloads if they're more profitable than the builder. Since we released v3.5.0 with that change, we've had two requests to add another flag that always prefers the builder.

The first of these requests was so that Rocket Pool (RP) can force some operators to always use a specific builder to ensure that the operator isn't secretly pocketing MEV.

The second was from @mcdee, who wants to always use the builder because:

> Vouch needs the blinded proposals from multiple beacon nodes to use the same execution payload so it can compare them, and also importantly so it can publish the fully unblinded block to all beacon nodes.

There may be views that we *shouldn't* add this feature to try and force more locally produced blocks. I see those views, but these two scenarios show to me that there's valid, Ethereum-aligned use-cases for this flag and I don't think we should stand in the way of our users.

Additionally, there's been scenarios on testnets where we've always wanted to use the builder (e.g., the "big blocks" tests). This flag would be useful there, too.

## Additional Info

Rather than an "always use the builder" flag, we could have a flag that's along the lines of "use the builder *unless* the value of the local builder is worth more than `x` ETH". I'm a little unsure about this complexity, though. I suspect that if you're going to be comparing payload values then you'll always want the most profitable one. The two use cases I listed above are "always use the builder" so I'm tempted to just stick with that.

**Comment by michaelsproul** [comment]
I agree it's clean to have an "always use the builder flag".

However there is also a nice symmetry in a profit margin flag like `--builder-profit-margin N` where `N` is allowed to be negative. The logic would be:

```rust
let block = if builder_value >= local_value + N {
    // use builder
    builder_block
} else {
   // use local block
   local_block
};
```

When `N` is positive it allows users who prefer local blocks to only take the builder block if it offers a substantial improvement (this was requested here: https://github.com/sigp/lighthouse/issues/3699#issuecomment-1426533691).

When `N` is negative it allows users who prefer builder blocks to only take the local block if it offers a substantial improvement. This could be used for the "always use the builder" case by setting `N=i128::MIN`.

There's no I256 type in ethereum-types, so we would have to use `i128` for the calculation. I think this isn't likely to be an issue as it still represents values from -3.4 x10^20 to +3.4 x10^20 ETH.

**Comment by paulhauner** [comment]
For whoever implements this, I believe we just need to add an extra bool to this `if` statement (and a flag to toggle it):

https://github.com/sigp/lighthouse/blob/0fb58a680d6f0c9f0dc8beecf142186debff9a8d/beacon_node/execution_layer/src/lib.rs#L851-L859

E.g.,

```rust
if !self.always_prefer_builder_payload && local_value >= relay_value {
   ...
}
```

I think we should be careful to always use the builder payload only *if it is available*. I.e., if the builder doesn't return a paylaod but the EL does, then we still want to use the EL payload. I don't think this it's reasonable or safe to skip a block when we can avoid it.

**Comment by divagant-martian** [comment]
@danielrachi would you be interested in this issue?

**Comment by danielrachi1** [comment]
Sure, I'll work on it during the weekend

**Comment by paulhauner** [comment]
Thanks @danielrachi, much appreciated!

We're planning a release for Wednesday the 8th and we need to get this included in that release. We need a day of testing before we can release too, so the timeline is a little tight on this one.

I know you're a voluntary contributor so I certainly don't want to rush you. If you feel like you can get it done in time, please do. Otherwise please let me know and we can take it ☺️

**Comment by danielrachi** [comment]
I'm almost done, I think. I just have to write some tests now :)

**Comment by danielrachi1** [comment]
@paulhauner I think it's almost ready, I just have to write more tests and solve a weird compilation error I'm having trouble with.

You can check the draft PR now, if you think I'm not in the right direction please feel free to take the issue, I don't want to cause any trouble with the release.

**Comment by michaelsproul** [comment]
@mcdee I'm curious, does Vouch also disable the builder health checks in order to work its magic with blinded blocks?

**Comment by mcdee** [comment]
It doesn't require them to be disabled, but if the beacon node decides not to use the payload from the relay Vouch will in turn reject the blinded block.  So if the beacon node is in a state where it thinks that it shouldn't be using relays:
1. Vouch requests a blinded block
1. the beacon node returns a blinded block with a locally-built payload
1. Vouch rejects the blinded block due to the transactions root not matching
1. Vouch requests a (non-blinded) block from the beacon node, and works with that

I have considered requesting that in step 2) the beacon node sends a 404 rather than building a blinded block with a local payload without telling us (it's a right pain on testnets in particular, where they can be unstable to the point that the circuit breaker is active a lot of the time) but understand why, in the general case, it makes sense to do this.

The only other thing worth noting is that if there are multiple beacon nodes with different circuit breaker conditions then as long as one of them is taking the payload from the relay we'll use that regardless of the relative value of the blocks in terms of consensus rewards, which is a shame but not much we can do about it.  So in that case it can make sense for the circuit breaker conditions to be rather lenient, if not entirely off.  It's down to the individual users who configure their beacon nodes as to how they want to set it, though.

**Comment by michaelsproul** [comment]
Thanks for the comprehensive reply!

We've got the always-prefer flag landing in v3.5.1 so it sounds like that should restore things to the previous status quo. It's interesting to think of minimally invasive improvements we could make on this front though, thanks again for the input.

**Comment by paulhauner** [comment]
Resolved via #4052 🎉
